testsuite/MDAnalysisTests/analysis/test_density.py

```diff
@@ -105,6 +113,78 @@ def test_check_set_unit_nolength(self, D):
         with pytest.raises(ValueError):
             D._check_set_unit(units)
 
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        assert D._check_set_unit(units) == None
+
+    def test_check_set_unit_density_none(self, D):
+        units = {'density': None}
+        D._check_set_unit(units)
+        del D.units['density']
+        D._check_set_unit({})
+        assert D.units['density'] == None
+
+    def test_parameters_isdensity(self, D):
+        with pytest.warns(UserWarning, match='Running make_density()'):
+            D.make_density()
+
+    def test_check_convert_density_grid_not_density(self, D1):
+        unit = 'nm^{-3}'
+        with pytest.raises(RuntimeError, match="The grid is not a density"):
+            D1.convert_density(unit)
+
+    def test_check_convert_density_value_error(self, D):
+        unit = 'A^{-2}'
+        with pytest.raises(ValueError, match="The name of the unit"):
+            D.convert_density(unit)
+
+    def test_check_convert_density_units_same_density_units(self, D):
+        unit = 'A^{-3}'
+        D_orig = copy.deepcopy(D)
+        D.convert_density(unit)
+        assert D.units['density'] == D_orig.units['density'] == unit
+        assert_almost_equal(D.grid, D_orig.grid)
+
+    def test_check_convert_density_units_density(self, D):
+        unit = 'nm^{-3}'
+        D_orig = copy.deepcopy(D)
+        D.convert_density(unit)
+        assert_almost_equal(D.grid, 10**3 * D_orig.grid)
+
+    def test_convert_length_same_length_units(self, D):
+        unit = 'A'
+        D_orig = copy.deepcopy(D)
+        D.convert_length(unit)
+        assert D.units['length'] == D_orig.units['length'] == unit
+        assert_almost_equal(D.grid, D_orig.grid)
+
+    def test_convert_length_other_length_units(self, D):
+        unit = 'nm'
+        D_orig = copy.deepcopy(D)
+        D.convert_length(unit)
+        assert D.units['length'] == unit
+        assert_almost_equal(D.grid, D_orig.grid)
+
+    def test_repr(self, D, D1):
+        D.__repr__()
+        D1.__repr__()
```

Comment:
str(D) should just call `__repr__` in the background right? I don't _think_ these two calls are needed?